usr/sbin/p11sak/p11sak.c

```diff
@@ -2537,19 +2749,62 @@ static CK_RV close_session(CK_SESSION_HANDLE session)
 
     rc = funcs->C_Logout(session);
     if (rc != CKR_OK) {
-        printf("Logout failed (error code 0x%lX: %s)\n", rc, p11_get_ckr(rc));
+        fprintf(stderr, "Logout failed (error code 0x%lX: %s)\n", rc, p11_get_ckr(rc));
         return rc;
     }
 
     rc = funcs->C_Finalize(NULL_PTR);
     if (rc != CKR_OK) {
-        printf("Error in C_Finalize: (error code 0x%lX: %s)\n", rc,
+        fprintf(stderr, "Error in C_Finalize: (error code 0x%lX: %s)\n", rc,
                 p11_get_ckr(rc));
     }
 
     return rc;
 }
 
+static CK_RV parse_file() {
+    FILE *fp;
+    char *file_loc = getenv("P11SAK_DEFAULT_CONF_FILE");
+    char *buf;
+    struct passwd *pw;
+
+    // open and parse file
+    if (file_loc != NULL) {
+        if ((fp = fopen(file_loc, "r")) == NULL) {
+            fprintf(stderr, "Failed to load %s: %s\n", file_loc, strerror(errno));
+            return CKR_ARGUMENTS_BAD;
+        }
+    } else {
+        pw = getpwuid(getuid());
+        if (pw == NULL) {
+            perror("Could not find home directory of current user!");
+            return CKR_CANCEL;
+        }
+        if (asprintf(&buf, "%s/p11sak_defined_attrs.conf", pw->pw_dir) == -1) {
```

Comment:
The file in the home folder is usually hidden.  So your format string should read `%s/.p11sak_defined_attrs.conf`.

Reply:
Done